compiler/luci/pass/src/PropagateConcatenationQparamPass.cpp

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2020 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/PropagateConcatenationQparamPass.h"
+
+#include "PropagateConcatenationQparamPassInternal.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/IR/AttrFusedActFunc.h>
+#include <luci/Log.h>
+
+#include <math.h>
+
+namespace
+{
+
+bool is_uint8_quantized(luci::CircleNode *node)
+{
+  if (node->dtype() == loco::DataType::U8 && node->quantparam() != nullptr &&
+      node->quantparam()->scale.size() > 0 && node->quantparam()->zerop.size() > 0)
+    return true;
+
+  return false;
+}
+
+void overwrite_quantparam(luci::CircleConcatenation *concat, luci::CircleNode *target)
+{
+  auto concat_qparam = concat->quantparam();
+  assert(concat_qparam != nullptr);
+
+  auto target_qparam = target->quantparam();
+  assert(target_qparam != nullptr);
+  target_qparam->min = concat_qparam->min;
+  target_qparam->max = concat_qparam->max;
+  target_qparam->scale = concat_qparam->scale;
+  target_qparam->zerop = concat_qparam->zerop;
+  target_qparam->quantized_dimension = concat_qparam->quantized_dimension;
+}
+
+} // namespace
+
+namespace luci
+{
+
+void propagate_concat_quantparam(luci::CircleConcatenation *concat)
+{
+  // Check if concat is uint8-quantized
+  if (!is_uint8_quantized(concat))
+    return;
+
+  // Check if concat has no fused activation function
+  if (concat->fusedActivationFunction() != luci::FusedActFunc::NONE)
+    return;
+
+  const auto num_inputs = concat->numValues();
+  for (uint32_t i = 0; i < num_inputs; i++)
+  {
+    auto node = loco::must_cast<luci::CircleNode *>(concat->arg(i));
+
+    // Skip if this input is CONCAT Op
+    if (node->opcode() == luci::CircleOpcode::CONCATENATION)
+      continue;
+
+    // Skip if this input is used by other Ops
+    auto succs = loco::succs(node);
+    if (succs.size() != 1)
+      continue;
+
+    assert(succs.find(concat) != succs.end());
+
+    // Quantize constant values
+    if (node->opcode() == luci::CircleOpcode::CIRCLECONST)
+    {
+      luci::CircleConst *const_node = loco::must_cast<luci::CircleConst *>(node);
+      assert(const_node->dtype() == loco::DataType::FLOAT32);
```

Comment:
Not sure I understand this.
If the concat has quantization parameters and U8 type, how can its input be float?

Reply:
QuantizeWithMinMaxPass quantizes output tensor of concat, while it does not quantize constant inputs of concat.
(Currently the only quantized constant tensors are weights and bias of conv, dw_conv, tconv, fc layers)

So when PropagateConcatenationQparamPass is called, the output tensor of concat is already quantized but const inputs of concat are not yet quantized.

Reply:
Do you mean that without this optimization the graph would be invalid?
I think that any transformation should leave the graph in valid state. This allows to make them independent on each other, and requires less "internal knowledge" of each transformation.

Reply:
Our quantizer does not have a general function to handle constant inputs of an Op (as mentioned above, we only handle constant inputs for conv, dw_conv, tconv, fc layer). This is because we need to quantize constant inputs in an operator-specific way. For example, ReduceMean can receive a constant input as `axis`, which should not be quantized. So I wanted to implement quantization logics for each operator as the need arises (ex: a new pattern occurs in target models).

> Do you mean that without this optimization the graph would be invalid?
> I think that any transformation should leave the graph in valid state. This allows to make them independent on each other, and requires less "internal knowledge" of each transformation.

Agreed. I think we have to throw an exception for the unsupported cases (const inputs of unsupported Ops) to prevent the graph from being invalid. I'll make another PR for that.

Reply:
> I'll make another PR for that.

👍 Do you want to make it first, or would you like to merge this one now?

> This is because we need to quantize constant inputs in an operator-specific way. For example, ReduceMean can receive a constant input as `axis`, which should not be quantized. 

I see. As a suggestion, I saw some kind of black- or white-listing function for that in TFLite, and it works for them well. I could try to find the exact place if you want.

Reply:
> Do you want to make it first, or would you like to merge this one now?

I'd like to merge this now because it seems to be a separate issue.

> I see. As a suggestion, I saw some kind of black- or white-listing function for that in TFLite, and it works for them well. I could try to find the exact place if you want.

It would be very helpful to specify the exact place for the code. Thanks :)

Reply:
@jinevening Here is the relevant [code](https://github.com/tensorflow/tensorflow/blob/312c11c4ecdc2c466895fc3be0a16d60c692422e/tensorflow/lite/tools/optimize/operator_property.cc#L67-L994) and [data structures](https://github.com/tensorflow/tensorflow/blob/312c11c4ecdc2c466895fc3be0a16d60c692422e/tensorflow/lite/tools/optimize/operator_property.h#L57-L126).
The code may look scary, but half of the function is just LSTM case.